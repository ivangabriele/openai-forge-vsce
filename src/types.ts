export type DocumentInfo = {
  absolutePath: string
  relativePath: string
  source: string
}

export enum ProjectFramework {
  REACT = 'React',
  TAURI = 'Tauri',
}

export enum ProjectLanguage {
  JAVASCRIPT = 'JavaScript',
  RUST = 'Rust',
  TYPESCRIPT = 'TypeScript',
}

export enum State {
  'RUNNING' = 'RUNNING',
  'STARTING' = 'STARTING',
  'STOPPED' = 'STOPPED',
  'STOPPING' = 'STOPPING',
}

export type WorkspaceInfo = {
  languages: ProjectLanguage[]
  mainFramework: ProjectFramework | undefined
  name: string
  rootPath: string
  subFrameworks: ProjectFramework[]
}

export enum WebSocketDataAction {
  ASK = 'ASK',
}

export type WebSocketData = {
  action: WebSocketDataAction
  message: string
}
