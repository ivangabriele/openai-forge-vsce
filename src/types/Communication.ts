export namespace Communication {
  export enum MessageAction {
    ASK = 'ASK',
    CONVEY = 'CONVEY',
    DECLARE = 'DECLARE',
    ORDER = 'ORDER',
  }

  export enum MessageMessage {
    AS_SATELLITE = 'AS_SATELLITE',
    PASS_MESSAGE = 'AS_SATELLITE',
    SET_CLIENTS_COUNT = 'SET_CLIENTS_COUNT',
    SET_ID = 'SET_ID',
  }

  export type Message =
    | {
        action: MessageAction.ASK
        message: string
      }
    | {
        $id: string | undefined
        action: MessageAction.CONVEY
        value: {
          action: MessageAction.ASK
          message: string
        }
      }
    | {
        action: MessageAction.DECLARE
        message: MessageMessage.AS_SATELLITE
      }
    | {
        action: MessageAction.ORDER
        message: MessageMessage.SET_CLIENTS_COUNT
        value: number
      }
    | {
        action: MessageAction.ORDER
        message: MessageMessage.SET_ID
        value: string
      }
}
