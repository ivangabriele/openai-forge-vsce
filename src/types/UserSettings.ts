import type { UserWorkspace } from './UserWorkspace'

export namespace UserSettings {
  export type Settings = {
    customEvaluators: CustomEvaluator[]
    prompt: {
      excludeProjectInfo: boolean
    }
  }

  export type CustomEvaluator = Omit<UserWorkspace.Evaluator, 'workingDirectoryAbsolutePath'> & {
    workingDirectoryRelativePath: string
  }
}
