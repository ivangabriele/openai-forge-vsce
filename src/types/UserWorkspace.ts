export namespace UserWorkspace {
  export type Info = {
    frameworks: Framework[]
    languages: Language[]
    name: string
    rootPath: string
    sourceCodeDocumentsLength: number
  }

  export type Evaluator = {
    command: string
    commandArgs: string[]
    extensions: string[]
    workingDirectoryAbsolutePath: string
  }

  export enum Framework {
    NEXT_JS = 'Next.js',
    REACT = 'React',
    TAURI = 'Tauri',
  }

  export enum Language {
    C_SHARP = 'C#',
    DART = 'Dart',
    ELIXIR = 'Elixir',
    GO = 'Go',
    HASKELL = 'Haskell',
    JAVA = 'Java',
    JAVASCRIPT = 'JavaScript',
    KOTLIN = 'Kotlin',
    PHP = 'PHP',
    PYTHON = 'Python',
    RUBY = 'Ruby',
    RUST = 'Rust',
    SCALA = 'Scala',
    TYPESCRIPT = 'TypeScript',
  }

  export type LanguagesStatistic = {
    count: Record<Language, number>
    ratio: Record<Language, number>
  }
}
