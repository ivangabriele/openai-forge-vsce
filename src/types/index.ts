import type { Promisable } from 'type-fest'

export enum MessageItemType {
  BUTTON = 'BUTTON',
  LINK = 'LINK',
}

export type MessageButton = {
  action: () => Promisable<void>
  label: string
  type: MessageItemType.BUTTON
}

export type MessageLink = {
  label: string
  type: MessageItemType.LINK
  url: string
}

export enum State {
  'FAILED' = 'FAILED',
  'RUNNING' = 'RUNNING',
  'STARTING' = 'STARTING',
  'STOPPED' = 'STOPPED',
  'STOPPING' = 'STOPPING',
}

export enum WebSocketDataAction {
  ASK = 'ASK',
}

export type WebSocketData = {
  action: WebSocketDataAction
  message: string
}

export { type UserSettings } from './UserSettings'
export { UserWorkspace } from './UserWorkspace'
