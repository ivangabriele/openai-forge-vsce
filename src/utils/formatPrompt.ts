import { workspace } from 'vscode'

import type { DocumentInfo, WorkspaceInfo } from '../types'

export function formatPrompt(workspaceInfo: WorkspaceInfo, documentInfos: DocumentInfo[]): string {
  const excludeProjectInfo = workspace.getConfiguration('openai-forge').get<boolean>('promt.excludeProjectInfo')

  return (
    !excludeProjectInfo
      ? [
          `Project:`,
          `- Name: ${workspaceInfo.name}`,
          ...(workspaceInfo.mainFramework
            ? [`- Framework: ${workspaceInfo.mainFramework}, ${workspaceInfo.subFrameworks.join(', ')}`]
            : []),
          `- Languages: ${workspaceInfo.languages.join(', ')}`,
          '',
        ]
      : []
  )
    .concat(
      documentInfos.reduce(
        (lines, documentInfo, index) => [
          ...lines,
          ...(index > 0 ? [''] : []),
          `${documentInfo.relativePath}:`,
          '```',
          documentInfo.source,
          '```',
        ],
        [] as string[],
      ),
    )
    .join('\n')
}
