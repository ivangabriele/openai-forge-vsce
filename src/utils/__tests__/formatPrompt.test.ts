import { ProjectFramework, ProjectLanguage, type WorkspaceInfo, type DocumentInfo } from '../../types'
import { formatPrompt } from '../formatPrompt'

describe('utils/formatPrompt()', () => {
  it('should return a string', () => {
    const project: WorkspaceInfo = {
      name: 'FooBar',
      languages: [ProjectLanguage.RUST, ProjectLanguage.TYPESCRIPT],
      subFrameworks: [ProjectFramework.REACT],
      mainFramework: ProjectFramework.TAURI,
      source: '',
      rootPath: '/home/user/acme/foobar',
    }

    const documents: DocumentInfo[] = [
      {
        absolutePath: '/home/user/acme/foobar',
        relativePath: 'src/index.tsx',
        source: 'Some TypeScript',
      },
      {
        absolutePath: '/home/user/acme/foobar',
        relativePath: 'src-tauri/src/main.rs',
        source: 'Some Rust',
      },
    ]

    const result = formatPrompt(project, documents)

    expect(result).toMatchSnapshot()
  })
})
