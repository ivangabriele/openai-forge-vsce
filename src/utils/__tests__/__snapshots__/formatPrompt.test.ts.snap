// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/formatPrompt() should return a string 1`] = `
"Project:
- Name: FooBar
- Framework: Tauri, React
- Languages: Rust, TypeScript

src/index.tsx:
\`\`\`
Some TypeScript
\`\`\`

src-tauri/src/main.rs:
\`\`\`
Some Rust
\`\`\`"
`;
